### ISSUE TITLE
Perf suggestion for crlf handling

### ISSUE THREAD

**Issue by jeffhostetler** [libgit2/libgit2#3212]
I haven't measured this but....

As long a the crlf filtering is loading the entire blob/file into a buffer, there are some opportunities here for perf gains.

In git_buf_text_is_binary() we examine the entire file.  While there we could count the number of CRLFs, and plain LFs and return those sums.  We should be able to use those sums in git_buf_text_gather_stats().  Or combine those 2 routines into 1.  Then in git_buf_text_lf_to_crlf() add this delta "(count_crlf - count_lf)" as padding to the amount of reserved space in the destination buffer so that the memcpy loop doesn't need to realloc as the file is being converted.

**Comment by ethomson** [comment]
Yeah, there's actually no need for us to call `git_buf_text_is_binary`, since we can instead capture the value directly out of `git_buf_text_lf_to_crlf`.

(This is part of what I did in #3199 - https://github.com/libgit2/libgit2/pull/3199/files#diff-35f50f22e19720adbf1cc9d960219642R246)

**Comment by whoisj** [comment]
Are there any plans to migrate the current CRLF filter to the newer stream filter model?

**Comment by ethomson** [comment]
I don't have any plans to, unless git.git limits the amount of a blob it reads to determine binaryness, there's likely no point.

https://github.com/git/git/blob/master/convert.c#L1304

**Comment by jeffhostetler** [comment]
So they're doing the conversion in memory too?   OK.  It does seem like we're doing something a little different because I can successfully stage the file using git.exe that I create in #3211.

Granted there's probably a boundary there where git.exe will fail too.  Perhaps I just hit ours.

**Comment by ethomson** [comment]
Maybe I'm misunderstanding their process.  I would like to be wrong here.

**Comment by jeffhostetler** [comment]
I'll dig a little further then.

**Comment by linquize** [comment]
Must be consistent with core-git

**Comment by carlosmn** [comment]
git does use streaming filters for crlf conversion (though I don't quite recall if it's in both directios). I remember this quite well, as it once had an infinite loop when the stars aligned due to not keeping enough state :disappointed: .

When staging a file, we can certainly take advantage of streaming data into libgit2 to reduce heap usage (though that comes with a bunch of extra disk IO since we won't know the final size). For getting data out of the odb, we don't have any streaming capabilities, so we already have to have the whole source data in memory, though streaming should help the peak usage, as we should be able to stream 64k (or whatever) chunks out to the workdir instead of keeping files around.

**Comment by ethomson** [comment]
The requirement that we look at every byte when deciding binaryness and EOL statistics makes me think that we need to either load the file into memory to do this check, or to have a rewindable or restartable stream.  This may only impact the to workdir direction, but this aspect makes steaming a challenge.

Obviously there may be ways to drive our overall memory usage down in crlf, or other clever approaches here.  (Since we are long lived, we can cache the "binaryness" or statistics of a blob indefinitely, barring reclamation for memory concerns.)

**Comment by jeffhostetler** [comment]
Just re-reading this thread. I forgot to ack in my first response that you addressed my initial is_binary suggestion in #3199.  My second concern (delta crlf vs lf) doesn't matter since I missed that you're padding the target by a generous amount.
